src/utils/ThreadPool.h

```diff
@@ -0,0 +1,384 @@
+#include "../Application.h" // accessing singleton's Application::thread_pool
+#include "../utils/math.h" // round_up_divide
+
+#include <cassert>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <deque>
+#include <functional> // std::function<>
+
+namespace cura
+{
+
+/*!
+ * \brief Very minimal and low level thread pool.
+ *
+ * Consider using `parallel_for()` instead, interfacing directly with this class should be reserved to concurrency primitives.
+ * ThreadPool can be described as a synchronized FIFO queue shared by a fleet of `std::thread`s.
+ * Tasks have the responsibility of unlocking the queue's lock passed as an argument while they do asynchronous work.
+ */
+class ThreadPool
+{
+  public:
+    using lock_t = std::unique_lock<std::mutex>;
+    using task_t = std::function<void(lock_t&)>;
+
+    //! Spawns a thread pool with `nthreads` threads
+    ThreadPool(size_t nthreads);
+
+    ~ThreadPool() { join(); }
+
+    //! Returns the number of threads
+    size_t thread_count() const { return threads.size(); }
+
+    //! Gets a lock on the queue, stopping the queuing or execution of new tasks while held
+    lock_t get_lock() { return lock_t(mutex); }
+
+    /*!
+     * \brief Pushes a new task while the queue is locked.
+     * \param func Closure that unlocks the local worker's lock passed as an argument while
+     * doing asynchronous work.
+     */
+    template<typename F>
+    void push(const lock_t& lock [[maybe_unused]], F&& func)
+    {
+        assert(lock);
+        tasks.push_back(std::forward<F>(func));
+        condition.notify_one();
+    }
+    /*!
+     * \brief Executes pending tasks while the predicates returns true
+     * This method doesn't wait unless predicate does (like implementation of ThreadPool::worker())
+     */
+    template<typename P> void work_while(lock_t& lock, P predicate)
+    {
+        assert(lock);
+        while(predicate() && !tasks.empty()) // Order is important: predicate() might wait on an empty queue
+        {
+            assert(!tasks.empty());
+            task_t task = std::move(tasks.front());
+            tasks.pop_front();
+
+            task(lock);
+            assert(lock);
+        }
+    }
+
+  private:
+    void worker();
+
+    void join();
+
+    std::mutex mutex;
+    std::condition_variable condition;
+    std::deque<task_t> tasks;
+    std::vector<std::thread> threads;
+    bool wait_for_new_tasks;
+};
+
+
+/// `std::make_signed_t` fails for non integral types in a way that doesn't allows SFINAE fallbacks. This alias solves that.
+template<typename T> using make_signed_if_integral_t = typename std::enable_if_t<std::is_integral_v<T>, std::make_signed<T>>::type;
+
+using std::distance;
+/// Overloads `std::distance()` to work on integral types
+template<typename Int, typename Signed=make_signed_if_integral_t<Int>>
+inline Signed distance(const Int& first, const Int& last)
+{
+    return static_cast<Signed>(last) - static_cast<Signed>(first);
+}
```

Comment:
I'm not sure where to put these declarations.
This enables the same implementation of `parallel_for` to work on integers and random access iterators.
Also, this could be replaced with `<` and `-`.